crates/apollo-compiler/src/validation/value.rs

```diff
@@ -0,0 +1,263 @@
+                        let is_undefined = !input_obj.fields().any(|f| f.name() == name.src());
+                        if is_undefined {
+                            return Some((name, value));
+                        }
+                        None
+                    });
+
+                    // Add a diagnostic if a value does not exist on the input
+                    // object type
+                    if let Some((name, value)) = undefined_field {
+                        diagnostics.push(
+                            ApolloDiagnostic::new(
+                                db,
+                                value.loc().into(),
+                                DiagnosticData::UndefinedValue {
+                                    value: name.src().into(),
+                                    definition: type_def.name().into(),
+                                },
+                            )
+                            .label(Label::new(
+                                value.loc(),
+                                format!("does not exist on `{}` type", type_def.name()),
+                            )),
+                        );
+                    }
+
+                    input_obj.fields().for_each(|f| {
+                        let ty = f.ty();
+                        let is_missing = !obj.iter().any(|(name, ..)| f.name() == name.src());
+
+                        // If no default value is provided and the input object
+                        // field’s type is non-null, an error should be raised
+                        if (ty.is_non_null() && f.default_value().is_none()) && is_missing {
+                            let mut diagnostic = ApolloDiagnostic::new(
+                                db,
+                                val.loc().into(),
+                                DiagnosticData::RequiredArgument {
+                                    name: f.name().into(),
+                                },
+                            );
+                            diagnostic = diagnostic.label(Label::new(
+                                val.loc(),
+                                format!("missing value for argument `{}`", f.name()),
+                            ));
+                            if let Some(loc) = ty.loc() {
+                                diagnostic =
+                                    diagnostic.label(Label::new(loc, "argument defined here"));
+                            }
+
+                            diagnostics.push(diagnostic)
+                        }
+
+                        obj.iter().for_each(|(name, v)| {
+                            let type_def = ty.type_def(db.upcast());
+                            if name.src() == f.name() {
+                                if let Some(_def) = type_def {
+                                    value_of_correct_type(db, ty, v, var_defs.clone(), diagnostics)
+                                }
+                            }
+                        })
```

Comment:
it looks like this `for_each` is only meant to actually do the work for 1 of the fields? maybe we can use `.find()` to get at the thing we need instead?

Reply:
we're already inside a loop, so `f` is a single field in the input object type here. there should be only one match unless the value has duplicate keys, which is already caught by other validations